[TT-16831] docs: add OpenTelemetry metrics pages and restructure metrics section (#1808)
TykTechnologies/tyk-docs: api-management/observability/guides/gateway-playbook.mdx

---
title: "Tyk Gateway Observability Playbook"
description: "Use native OTLP metrics to monitor Tyk Gateway RED signals, diagnose common failures, and configure Prometheus alerting rules."
keywords: "opentelemetry, metrics, prometheus, alerting, latency, error rate, goroutines, observability, RED metrics, response flags"
sidebarTitle: "Gateway Observability Playbook"
---

## Overview

Tyk Gateway 5.13+ exports metrics natively via OTLP, removing the need for Tyk Pump as an intermediary for Prometheus-based observability:

> ```

Reviewer: Definitely needs a better diagram from Jen

Gateway → OTel Collector → any OTLP backend
                ↘
                 Prometheus scrape endpoint
```

> Every request the gateway handles generates three signal types that share common identifiers, enabling end-to-end correlation:

Reviewer: ```suggestion
Every request the Gateway handles generates three signal types that share common identifiers, enabling end-to-end correlation:
```

| Identifier | In logs | In metrics (Prometheus label) | In traces |
|-----------|---------|-------------------------------|-----------|
| API ID | `api_id` | `tyk_api_id` | `tyk.api.id` attribute |
| Response flag | `response_flag` | `tyk_response_flag` | — |
| Consumer key | `api_key` | Available via custom `api_metrics` config | `tyk.api.apikey` |
| Trace ID | `trace_id` | — (use exemplars) | span `traceId` |
| Gateway instance | — | `service_name` | `service.instance.id` |

This guide covers two categories of metrics:

- **RED metrics**: What your APIs are doing right now: request rate, error classification, and latency decomposition.

> - **Gateway health metrics**: What the gateway process itself is doing: memory, goroutines, and configuration state.

Reviewer: ```suggestion
- **Gateway health metrics**: What the Gateway process itself is doing: memory, goroutines, and configuration state.
```

## Prerequisites

- Tyk Gateway 5.13 or later
- OpenTelemetry metrics enabled (`opentelemetry.metrics.enabled: true` in `tyk.conf`, or `TYK_GW_OPENTELEMETRY_METRICS_ENABLED=true`)
- An OTel Collector configured to export to Prometheus (or a compatible OTLP backend)
- Prometheus scraping the Collector's metrics endpoint

If you haven't set up the OTel pipeline yet, see [OpenTelemetry tracing and metrics](/api-management/traces) and the [OpenTelemetry demo guide](/api-management/opentelemetry-demo).

## Key thresholds at a glance

<Note>
Use these as starting points and adjust based on your API's traffic profile and SLOs.

| Signal | Healthy threshold | Alert threshold |
|--------|------------------|-----------------|
| p95 end-to-end latency | < 500ms | > 1s for 5 min |
| p99 end-to-end latency | < 1s | > 2s for 5 min |
| Error rate (non-2xx) | < 2% | > 10% for 5 min |
| Gateway-only avg latency | < 10ms | > 50ms |
| Goroutine count | < 2,000 | > 5,000 for 10 min |
| Auth failure rate | < 0.01/s baseline | > 0.1/s for 2 min |
| Upstream error rate (URS) | 0 | Any sustained for 3 min |
</Note>

---

## Core RED Metrics

### Request Rate

> Request rate tells you how much traffic the gateway is handling. Sudden drops indicate that requests are not reaching the gateway: DNS failures, network partitions between clients and the gateway, or client-side misconfiguration.

Reviewer: ```suggestion
Request rate tells you how much traffic the Gateway is handling. Sudden drops indicate that requests are not reaching the Gateway: DNS failures, network partitions between clients and the gateway, or client-side misconfiguration.
```

**Metrics:**

| Metric name (OTel) | Prometheus name | Type | Key dimensions |
|-------------------|-----------------|------|----------------|
| `tyk.api.requests.total` | `tyk_api_requests_total` | counter | `http_request_method`, `http_response_status_code`, `tyk_api_id` |

**What are normal request rate thresholds?**

There is no universal baseline. Request rate varies by deployment. Establish a baseline over 7 days and alert on deviations greater than ±30% from the rolling average for the same hour of the prior week.

A more actionable check is to watch for a sudden drop to near-zero on a previously active API:

```promql
# Alert if rate drops > 80% compared to 1 hour ago
rate(tyk_api_requests_total{tyk_api_id="<api_id>"}[5m])
  < 0.2 * rate(tyk_api_requests_total{tyk_api_id="<api_id>"}[5m] offset 1h)
```

**Troubleshooting unexpected changes in request rate:**

| Issue | Possible Causes | Remediation |
|-------|----------------|-------------|

> | Rate drops to zero on one API, others normal | API definition deleted or disabled; DNS for that API's listen path changed | Check gateway logs for `api_id` config errors; verify API is active in Tyk Dashboard |

Reviewer: ```suggestion
| Rate drops to zero on one API, others normal | API definition deleted or disabled; DNS for that API's listen path changed | Check Gateway logs for `api_id` config errors; verify API is active in Tyk Dashboard |
```

> | Rate drops uniformly across all APIs | Network partition between clients and gateway; load balancer health check failure; OTel Collector not receiving (metrics gap, not a real traffic drop) | Check gateway health endpoint; confirm traffic is actually dropping and not just metric export failing |

Reviewer: ```suggestion
| Rate drops uniformly across all APIs | Network partition between clients and Gateway; load balancer health check failure; OTel Collector not receiving (metrics gap, not a real traffic drop) | Check Gateway health endpoint; confirm traffic is actually dropping and not just metric export failing |
```

| Rate spikes suddenly on one API | Traffic surge, DDoS, or runaway batch client | Filter logs by `api_id` and group by `api_key` to identify the calling consumer |

> | Rate split unevenly across gateway instances | Sticky sessions or uneven load balancer weights | Check `service.instance.id` label in metrics to compare per-instance rates |

Reviewer: ```suggestion
| Rate split unevenly across Gateway instances | Sticky sessions or uneven load balancer weights | Check `service.instance.id` label in metrics to compare per-instance rates |
```

| *(HTTP status code, e.g. `200`)* | 200 | **Yes** | *(absent)* | Successful request — `tyk_response_flag` is set to the HTTP status code |
| `AMF` | 401 | **No** | `AuthKey` | Auth header entirely absent |
| `AKI` | 403 | **No** | `AuthKey` | Auth header present, key invalid or expired |
| `QEX` | 403 | **No** | `RateLimitAndQuotaCheck` | Key's quota window exhausted |
| `RLT` | 429 | **No** | `RateLimitAndQuotaCheck` | Per-second rate limit exceeded |
| `URS` | 500 | **Yes** | `Upstream` | Upstream returned a 5xx error |

**What error rate thresholds should I set?**

A non-2xx rate below **2%** is a healthy baseline for most APIs with authenticated consumers. Alert at **10%** over a 5-minute window for most APIs. For payment or health endpoints, tighten to 1%.

Calculate the current error rate:

```promql
(
  rate(tyk_api_requests_total{http_response_status_code!~"2..", tyk_api_id="<api_id>"}[5m])
  /
  rate(tyk_api_requests_total{tyk_api_id="<api_id>"}[5m])
) * 100
```

Classify errors by flag to route to the right runbook:

```promql
sum by (tyk_response_flag) (
  rate(http_server_request_duration_seconds_count{tyk_api_id="<api_id>"}[5m])
)
```

<Note>

> Check `upstream_latency` in logs first. If it is `0`, the request never left the gateway: the error originated in auth, quota, or rate-limit middleware. If `upstream_latency > 0` and `response_flag = URS`, the upstream itself failed.

Reviewer: ```suggestion
Check `upstream_latency` in logs first. If it is `0`, the request never left the Gateway: the error originated in auth, quota, or rate-limit middleware. If `upstream_latency > 0` and `response_flag = URS`, the upstream itself failed.
```

</Note>

**Troubleshooting elevated error rates:**

| Issue | Possible Causes | Remediation |
|-------|----------------|-------------|

> | Surge in `AMF` (401) | Auth header entirely absent at Tyk — the client never sends it, or an intermediary (load balancer, CDN, reverse proxy) strips it before reaching the gateway | Identify source: `{prefix="access-log"} \| json \| response_flag="AMF" \| line_format "ip={{.client_ip}} api={{.api_name}}"`. Single IP after a network change → suspect header stripping by an intermediary. Wide IP spread → clients calling the wrong listen path |

Reviewer: ```suggestion
| Surge in `AMF` (401) | Auth header entirely absent at Tyk — the client never sends it, or an intermediary (load balancer, CDN, reverse proxy) strips it before reaching the Gateway | Identify source: `{prefix="access-log"} \| json \| response_flag="AMF" \| line_format "ip={{.client_ip}} api={{.api_name}}"`. Single IP after a network change → suspect header stripping by an intermediary. Wide IP spread → clients calling the wrong listen path |
```

| Surge in `AKI` (403) | Key was rotated, expired, or revoked; credential stuffing attack | Single source IP → contact consumer. Wide IP spread → credential attack; tighten rate limits |
| Sustained `QEX` (403) | Consumer legitimately exhausted their quota tier | Identify consumer via `key` field in logs; invite upgrade or raise quota ceiling in policy |
| `RLT` (429) climbing | Legitimate traffic burst; retry storm after upstream error | Check whether RLT is protecting a degraded upstream (correct) or blocking legitimate traffic (adjust limit). Add backoff guidance to consumer |
| Sustained `URS` (500) | Backend service degraded; upstream 5xx errors | Extract `error_target` (upstream hostname) and `upstream_addr` from logs; escalate to backend team. Check retry/circuit-breaker plugin config |

> | Non-2xx with no `response_flag` | Auth plugin returning custom status; unhandled middleware error | Search gateway error logs for the request's `trace_id` |

Reviewer: ```suggestion
| Non-2xx with no `response_flag` | Auth plugin returning custom status; unhandled middleware error | Search Gateway error logs for the request's `trace_id` |
```

---

### Latency

> Tyk exports three latency histograms that let you decompose end-to-end response time into what the gateway spent versus what the upstream spent. Use them together. Any one alone gives an incomplete picture.

Reviewer: ```suggestion
Tyk exports three latency histograms that let you decompose end-to-end response time into what the Gateway spent versus what the upstream spent. Use them together. Any one alone gives an incomplete picture.
```

**Metrics:**

| Prometheus name | Measures | Key labels |
|----------------|---------|------------|
| `http_server_request_duration_seconds` | Total time from first byte received to last byte sent (client view) | `http_request_method`, `http_response_status_code`, `tyk_api_id`, `tyk_response_flag` |
| `tyk_gateway_request_duration_seconds` | Gateway-only processing time (middleware chain, auth, rate limiting) | `http_request_method`, `tyk_api_id`, `tyk_response_flag` |
| `tyk_upstream_request_duration_seconds` | Time waiting for the upstream service to respond | `http_request_method`, `tyk_api_id`, `tyk_response_flag` |

**Histogram bucket boundaries (all three metrics, in seconds):**

```
0.001  0.005  0.01  0.025  0.05  0.1  0.25  0.5  1  2.5  5  10  +Inf
```

**What latency thresholds should I set?**

| Percentile | Target | Alert |
|-----------|--------|-------|
| p50 (median) | < 100ms | — |
| p95 | < 500ms | > 1s for 5 min |
| p99 | < 1s | > 2s for 5 min |
| Gateway-only average | < 10ms | > 50ms |

<Note>

> **Latency isolation rule**: If `http_server_request_duration_seconds` p95 is high but `tyk_gateway_request_duration_seconds` p95 is < 10ms, **the gateway is healthy**: the latency is upstream. If both histograms are elevated, the gateway is the bottleneck.

Reviewer: ```suggestion
**Latency isolation rule**: If `http_server_request_duration_seconds` p95 is high but `tyk_gateway_request_duration_seconds` p95 is < 10ms, **the Gateway is healthy**: the latency is upstream. If both histograms are elevated, the gateway is the bottleneck.
```

</Note>

Query p95 and p99 end-to-end:

```promql
histogram_quantile(0.95, rate(http_server_request_duration_seconds_bucket{tyk_api_id="<api_id>"}[5m]))
histogram_quantile(0.99, rate(http_server_request_duration_seconds_bucket{tyk_api_id="<api_id>"}[5m]))
```

> Query gateway-only average:

Reviewer: ```suggestion
Query Gateway-only average:
```

```promql
rate(tyk_gateway_request_duration_seconds_sum{tyk_api_id="<api_id>"}[5m])
  / rate(tyk_gateway_request_duration_seconds_count{tyk_api_id="<api_id>"}[5m])
```

**Troubleshooting high latency:**

| Issue | Possible Causes | Remediation |
|-------|----------------|-------------|
| High p95 overall, gateway avg < 10ms | Upstream service degradation; slow backend endpoint | Check `tyk_upstream_request_duration_seconds` p95. Extract `error_target` from URS logs. Escalate to backend team |
| Both gateway and upstream histograms elevated | Gateway CPU saturation; goroutine backpressure | Check `go_goroutine_count` trend and `go_memory_used_bytes`. See [Goroutines](#goroutines) below |
| p99 widening without p50 or p95 change | Load-induced queuing tail; a fraction of requests hitting a slow upstream path | Filter slow requests in logs: `latency_total > 1000`. Copy `trace_id` → Jaeger/Tempo to see which span is slow |

> | High latency scoped to one consumer | Consumer calling a slow upstream endpoint (not a gateway issue) | Per-key filter in logs: `api_id="<api_id>" \| line_format "key={{.api_key}} latency={{.latency_total}} path={{.path}}"` |

Reviewer: ```suggestion
| High latency scoped to one client | Client calling a slow upstream endpoint (not a gateway issue) | Per-key filter in logs: `api_id="<api_id>" \| line_format "key={{.api_key}} latency={{.latency_total}} path={{.path}}"` |
```

| Latency spike after config reload | Cold cache or policy recalculation during reload | Check `tyk_gateway_config_reload_duration_seconds`. Latency spike should resolve within 1–2 minutes |

---

## Gateway Health Metrics

> Gateway health metrics reflect the internal state of the gateway process itself, independent of API traffic. They are available when `opentelemetry.metrics.enabled: true` (Tyk Gateway 5.13+) and can be disabled independently with `runtime_metrics: false` if you only need RED signals.

Reviewer: ```suggestion
Gateway health metrics reflect the internal state of the Gateway process itself, independent of API traffic. They are available when `opentelemetry.metrics.enabled: true` (Tyk Gateway 5.13+) and can be disabled independently with `runtime_metrics: false` if you only need RED signals.
```

| Memory growing proportionally with API count | Normal — each API definition has memory overhead | Increase instance memory; review whether all loaded APIs are still needed |
| Memory growing faster than expected with stable API count | High allocation pressure from request handling | Check rate of `go_memory_allocations_total`; if climbing steeply, contact Tyk support with a heap profile |

---

### Goroutines

**Metrics:**

| Metric name (OTel) | Prometheus name | Type | What it tells you |
|-------------------|-----------------|------|------------------|
| `go.goroutine.count` | `go_goroutine_count` | Gauge | Number of active goroutines. Monotonic growth = leak |
| `go.processor.limit` | `go_processor_limit` | Gauge | Number of OS threads available (GOMAXPROCS) |
| `go.config.gogc` | `go_config_gogc_percent` | Gauge | GC target percentage (GOGC env var) |

**What goroutine thresholds should I set?**

> A healthy gateway at moderate load runs with 500–2,000 goroutines. Alert at **5,000 goroutines sustained for 10 minutes**. A one-time spike during a traffic burst is normal; a monotonically increasing trend over hours is not.

Reviewer: ```suggestion
A healthy Gateway at moderate load runs with 500–2,000 goroutines. Alert at **5,000 goroutines sustained for 10 minutes**. A one-time spike during a traffic burst is normal; a monotonically increasing trend over hours is not.
```

**Identify:**

```promql
rate(http_server_request_duration_seconds_count{tyk_response_flag=~"AMF|AKI"}[5m])
```

In Loki:

```logql
{prefix="access-log"} | json | response_flag=~"AMF|AKI"
  | line_format "{{.time}} {{.response_flag}} ip={{.client_ip}} api={{.api_name}}"
```

**Classify:**

| Pattern | Likely cause | Action |
|---------|-------------|--------|
| Single source IP, sustained after deployment | Credential rotation missed in deploy config | Contact consumer; verify keys still valid in Tyk Dashboard |
| Single source IP, random key attempts | Misconfigured integration (wrong env endpoint) | Contact consumer |
| Wide IP spread, varied keys | Credential stuffing or API scanning | Add IP allowlisting; tighten rate limits |
| All APIs, after Tyk Dashboard restart | Gateway missed key sync | Trigger manual key sync; check Dashboard connectivity |

<Note>
`AMF` and `AKI` are identical from a consumer-impact perspective: both result in the request never reaching the upstream. The distinction matters for root cause: `AMF` means the client didn't send a key at all; `AKI` means the client sent a key that Tyk cannot resolve.
</Note>

---

### Cardinality Overflow

> The gateway caps at **2,000 unique attribute combinations per instrument** by default (`cardinality_limit` in `tyk.conf`). When this cap is reached, additional data points are recorded with `otel.metric.overflow = true` rather than creating new time series.

Reviewer: ```suggestion
The Gateway caps at **2,000 unique attribute combinations per instrument** by default (see [cardinality control](/api-management/logs-metrics#cardinality-control)). When this cap is reached, additional data points are recorded with `otel.metric.overflow = true` rather than creating new time series.
```

**Detect:**

```promql
# Any non-zero result means cardinality overflow is occurring
tyk_api_requests_total{otel_metric_overflow="true"}
```

**Troubleshoot:**

| Issue | Possible Causes | Remediation |
|-------|----------------|-------------|
| Overflow on `tyk_api_requests_total` | More than 2,000 unique `api_id × method × status_code` combinations | Audit which APIs are generating unusual method/status combinations; consider raising `cardinality_limit` — contact Tyk support before doing so in production |
| Overflow after adding custom dimensions | Custom dimension like `client_ip` or `user_id` is high-cardinality | Remove the high-cardinality dimension or scope it with a hash/prefix |

<Note>

> Cardinality overflow does **not** drop data. The aggregate counts are preserved in the overflow bucket. You will see correct totals but lose the ability to break down data by the overflowing dimension combination.

Reviewer: ```suggestion
Cardinality overflow does **not** drop data. The aggregate counts are preserved in the overflow bucket. You will see correct totals but lose the ability to break the data down by the overflowing dimension combination.
```

</Note>

---

### Retry Storm

> Client retries without exponential backoff amplify an upstream failure. `RLT` (429) responses trigger more retries, which hit rate limits, which trigger more retries, a self-reinforcing loop that increases load on both the gateway and the upstream.

Reviewer: ```suggestion
Client retries without exponential backoff amplify an upstream failure. `RLT` (429) responses trigger more retries, which hit rate limits, which trigger more retries, a self-reinforcing loop that increases load on both the Gateway and the upstream.
```

**Identify:**

A retry storm shows `RLT` rate climbing while overall request rate is also climbing:

```promql
# RLT rate climbing
rate(http_server_request_duration_seconds_count{tyk_response_flag="RLT"}[1m])

# If this is also climbing, clients are retrying
rate(tyk_api_requests_total[1m])
```

**Troubleshoot:**

| Issue | Possible Causes | Remediation |
|-------|----------------|-------------|
| RLT climbing with overall rate climbing | Clients retrying 429s without backoff | Identify consumer via logs: `response_flag="RLT" \| line_format "key={{.key}} ip={{.client_ip}}"`. Advise consumer to implement exponential backoff with jitter |
| RLT starts immediately after upstream error (URS) | Upstream degradation triggers client retries which hit rate limits | Fix the upstream first. The rate limit is correctly protecting the backend during degradation |
| RLT on a single consumer, others unaffected | Single consumer batch job sending bursts | Work with consumer to spread requests or raise their rate limit ceiling |

---

## Set Up Alerting

> Prometheus Alertmanager handles alert routing. The gateway emits Prometheus-compatible metrics via the OTel Collector; alert rules evaluate against those metrics.

Reviewer: ```suggestion
Prometheus Alertmanager handles alert routing. The Gateway emits Prometheus-compatible metrics via the OTel Collector; alert rules evaluate against those metrics.
```

## Gateway health

# Goroutine count trend
go_goroutine_count{service_name="tyk-gateway"}

# Memory in use (go_memory_type="other" = heap-adjacent; use for leak detection)
go_memory_used_bytes{service_name="tyk-gateway", go_memory_type="other"}

# GC target heap size
go_memory_gc_goal_bytes{service_name="tyk-gateway"}

# Configured GOMEMLIMIT (denominator for utilization alert)
go_memory_limit_bytes{service_name="tyk-gateway"}

# APIs and policies loaded
tyk_gateway_apis_loaded{service_name="tyk-gateway"}
tyk_gateway_policies_loaded{service_name="tyk-gateway"}

## Consumer-level breakdown (requires custom api_metrics configuration)
# These metrics are NOT emitted by default. They must be defined in
# opentelemetry.metrics.api_metrics in tyk.conf.

# Example: requests by API key (last 6 chars), if configured with api_key session dimension
# increase(tyk_requests_by_apikey_total{tyk_api_id="<api_id>"}[1h]) by (api_key)

# Example: 5xx errors by route, if configured with endpoint metadata dimension
# increase(tyk_requests_by_route_total{tyk_api_id="<api_id>", http_response_status_code="500"}[1h]) by (tyk_endpoint)
```

---

## Next Steps

> 1. **Enable the OTel pipeline**: if you haven't yet, follow the setup in [OpenTelemetry in Tyk Gateway](/api-management/observability) to enable native OTLP export and route signals to your observability backend.

Reviewer: ```suggestion
1. **Enable the OTel pipeline**: if you haven't yet, follow the [setup instructions](/api-management/observability) to enable native OTLP export and route signals to your observability backend.
```
